### PR TITLE
fix: memory leak in @elastic/elasticsearch instrumentation

### PR DESCRIPTION
**Pull request by trentm**
tl;dr: Use a WeakMap as a WeakMap, not as an Object.
This leak was introduced in #2484.

Fixes: #2569

### Checklist

- [x] Implement code
- [x] Add CHANGELOG.asciidoc entry
- [x] Commit message follows [commit guidelines](https://github.com/elastic/apm-agent-nodejs/blob/main/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

File: lib/instrumentation/modules/@elastic/elasticsearch.js [hunk 1]
```diff
@@ -70,7 +70,7 @@ module.exports = function (elasticsearch, agent, { version, enabled }) {
           if (result) {
             const currSpan = ins.currSpan()
             if (currSpan) {
-              diagResultFromSpan[currSpan] = result
+              diagResultFromSpan.set(currSpan, result)
```
Comment (trentm) [inline lib/instrumentation/modules/@elastic/elasticsearch.js:73]: :embarrassment:

**Comment by trentm** [comment]
run module tests for @elastic/elasticsearch,@elastic/elasticsearch-canary

File: lib/instrumentation/modules/@elastic/elasticsearch.js (hunk 1 above)
Comment (astorm) [reply lib/instrumentation/modules/@elastic/elasticsearch.js:73]: it got by my review too so sign me up for the :embarrassment: newsletter.

**Review by astorm — APPROVED** [review]
approved

File: lib/instrumentation/modules/@elastic/elasticsearch.js (hunk 1 above)
Comment (felixbarny) [reply lib/instrumentation/modules/@elastic/elasticsearch.js:73]: Wow, that's a pitfall if I have ever seen one.

Comment (trentm) [reply lib/instrumentation/modules/@elastic/elasticsearch.js:73]: Dynamic languages for you. Not sure if TypeScript would have caught this for me.

**Pull request merged**